extern/boostd-data/bench/postgres.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	return &Postgres{defDb: defDb, connectString: connectString}, nil
+}
+
+func (db *Postgres) Name() string {
+	return "Postgres DB"
+}
+
+//go:embed create_tables.sql
+var createTables string
+
+func (db *Postgres) Init(ctx context.Context) error {
+	// Drop the db in case it didn't get cleaned up last time the program ran
+	_, _ = db.defDb.ExecContext(ctx, `DROP database bench`)
+
+	_, err := db.defDb.ExecContext(ctx, `CREATE DATABASE bench`)
+	if err != nil {
+		return fmt.Errorf("creating database bench: %w", err)
+	}
+
+	err = db.connect(ctx)
+	if err != nil {
+		return fmt.Errorf("connecting to db: %w", err)
+	}
+
+	_, err = db.db.ExecContext(ctx, createTables)
+	if err != nil {
+		return fmt.Errorf("creating tables: %w", err)
+	}
+
+	return nil
+}
+
+func (db *Postgres) connect(ctx context.Context) error {
+	benchConnStr, err := getConnStringWithDb(db.connectString, "bench")
+	if err != nil {
+		return err
+	}
+	benchDb, err := sql.Open("postgres", benchConnStr)
+	if err != nil {
+		return fmt.Errorf("connecting to database Bench: %w", err)
+	}
+	db.db = benchDb
+
+	return nil
+}
+
+func (db *Postgres) Cleanup(ctx context.Context) error {
+	err := db.db.Close()
+	if err != nil {
+		return err
+	}
+
+	_, err = db.defDb.ExecContext(ctx, `DROP database bench`)
+	return err
+}
+
+func (db *Postgres) GetBlockSample(ctx context.Context, count int) ([]pieceBlock, error) {
+	qry := `SELECT PieceCid, PayloadMultihash FROM PieceBlockOffsetSize ORDER BY RANDOM() LIMIT $1`
+	rows, err := db.db.QueryContext(ctx, qry, count)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	pbs := make([]pieceBlock, 0, count)
+	for rows.Next() {
+		var pieceCidBz, payloadMHBz []byte
+		err := rows.Scan(&pieceCidBz, &payloadMHBz)
+		if err != nil {
+			return nil, fmt.Errorf("scanning row: %w", err)
+		}
+
+		_, pcid, err := cid.CidFromBytes(pieceCidBz)
+		if err != nil {
+			return nil, fmt.Errorf("scanning piece cid: %w", err)
+		}
+		_, pmh, err := multihash.MHFromBytes(payloadMHBz)
+		if err != nil {
+			return nil, fmt.Errorf("scanning mulithash: %w", err)
+		}
+
+		pbs = append(pbs, pieceBlock{
+			PieceCid:         pcid,
+			PayloadMultihash: pmh,
+		})
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+
+	return pbs, nil
+}
+
+func (db *Postgres) AddIndexRecords(ctx context.Context, pieceCid cid.Cid, recs []model.Record) error {
+	if len(recs) == 0 {
+		return nil
+	}
+
+	tx, err := db.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Commit()
+
+	// Add payload to pieces index
+	vals := ""
+	args := make([]interface{}, 0, len(recs)*2)
+	for i, rec := range recs {
+		if i > 0 {
+			vals = vals + ","
+		}
+		vals = vals + fmt.Sprintf("($%d,$%d)", (i*2)+1, (i*2)+2)
+		args = append(args, rec.Cid.Hash(), pieceCid.Bytes())
+	}
+	_, err = tx.ExecContext(ctx, `INSERT INTO PayloadToPieces (PayloadMultihash, PieceCids) VALUES `+vals, args...)
+	if err != nil {
+		return fmt.Errorf("executing insert: %w", err)
+	}
```

Comment:
This is not the fastest way to go about doing large inserts to postgres, especially if you have some latency.

In my experience the fastest way, by far, is `copy .. from stdin` into a temp table, then insert-select from that table ignoring conflicts.

See e.g. https://github.com/filecoin-project/lotus/blob/v1.4.2/cmd/lotus-chainwatch/processor/messages.go#L121-L159